source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -98,15 +101,28 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
   return status;
 }
 
-Http::FilterDataStatus Filter::onData(ProcessorState& state, ProcessingMode::BodySendMode body_mode,
-                                      Buffer::Instance& data, bool end_stream) {
-  if (state.callbackState() == ProcessorState::CallbackState::Headers) {
+Http::FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data,
+                                      bool end_stream) {
+  if (end_stream) {
+    state.setCompleteBodyDelivered(true);
+  }
+
+  bool send_trailers = false;
+  Http::HeaderMap* new_trailers = nullptr;
+  if (end_stream && state.sendTrailers()) {
+    // We're at the end of the stream, but the filter wants to process trailers.
+    // We need to add them right here so we can process them later.
+    ENVOY_LOG(trace, "Creating new, empty trailers");
+    new_trailers = state.addTrailers();
+    state.setTrailersDelivered(true);
```

Comment:
What happens if we bail early in a `return` below? It's a bit hard to follow if we're definitely going to send.

Reply:
I updated the code to make the comments clearer, and renamed the "send_trailers" flag to make it easier to understand I hope. The logic is still correct -- we need to add the trailers here now because at least according to the comments in filter.h, we can only do it here, even if we're going to return right away and set the headers later.

Reply:
I guess I still find it a bit misleading having `setTrailersDelivered(true)` set here when they haven't yet been delivered. Is there a way to make this clearer in the state object?

Reply:
What if instead of "completeBodyDelivered" and "trailersDelivered" flags in that state object, we rename them to "completeBodyAvailable" and "trailersAvailable." That would show that those flags are related to what happened in the filter state, and not what the actual filter has done so far. Does that make it clearer? If so I'll make the change.

Reply:
SGTM

Reply:
Done in the latest commit.